--- qt-configure-module.orig	2024-01-03 12:34:38.268021696 +0100
+++ qt-configure-module	2024-01-03 12:54:50.541363644 +0100
@@ -33,6 +33,6 @@
 echo "$arg" >> "$optfile"
 done
 
-cmake_script_path="$script_dir_path/..\lib\cmake\Qt6/QtProcessConfigureArgs.cmake"
+cmake_script_path="$script_dir_path/../lib/cmake/Qt6/QtProcessConfigureArgs.cmake"
 qt_cmake_private_path="$script_dir_path/../libexec"
 "$qt_cmake_private_path/qt-cmake-private" -DOPTFILE=$optfile -DMODULE_ROOT="$module_root" -DCMAKE_COMMAND="$qt_cmake_private_path/qt-cmake-private" -P "$cmake_script_path"
