--- Qt6Dependencies.cmake.orig	2023-08-28 07:28:07.630920804 +0000
+++ Qt6Dependencies.cmake	2023-08-28 07:29:08.806970500 +0000
@@ -1,8 +1,8 @@
 set(Qt6_FOUND FALSE)
 
 set(__qt_platform_requires_host_info_package "TRUE")
-set(__qt_platform_initial_qt_host_path "/Users/qt/work/install")
-set(__qt_platform_initial_qt_host_path_cmake_dir "/Users/qt/work/install/lib/cmake")
+set(__qt_platform_initial_qt_host_path "/opt/hostedtoolcache/Qt/6.5.2/gcc_64")
+set(__qt_platform_initial_qt_host_path_cmake_dir "/opt/hostedtoolcache/Qt/6.5.2/gcc_64/lib/cmake")
 
 _qt_internal_setup_qt_host_path(
     "${__qt_platform_requires_host_info_package}"
