--- QtBuildInternalsExtra.cmake.orig	2023-08-28 07:31:02.027063145 +0000
+++ QtBuildInternalsExtra.cmake	2023-08-28 07:31:57.043108429 +0000
@@ -42,8 +42,8 @@
 if(CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT AND
         NOT QT_BUILD_INTERNALS_NO_FORCE_SET_INSTALL_PREFIX
         AND NOT QT_SUPERBUILD)
-    set(qtbi_orig_prefix "C:/Qt/Qt-6.5.2")
-    set(qtbi_orig_staging_prefix "/Users/qt/work/install/target")
+    set(qtbi_orig_prefix "/opt/hostedtoolcache/Qt")
+    set(qtbi_orig_staging_prefix "/opt/hostedtoolcache/Qt/6.5.2/wasm_singlethread")
     qt_internal_new_prefix(qtbi_new_prefix
         "${QT_BUILD_INTERNALS_RELOCATABLE_INSTALL_PREFIX}"
         "${qtbi_orig_prefix}")
